### ISSUE TITLE
What if zoom is 0 and 32?

### ISSUE THREAD

**Issue by jingsam** [mapbox/tippecanoe#677]
In epsg3857totile function https://github.com/mapbox/tippecanoe/blob/master/projection.cpp#L77-L83 :
```
	*x = ix * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31);
	*y = ((1LL << 32) - 1) - (iy * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31));

	if (zoom != 0) {
		*x >>= (32 - zoom);
		*y >>= (32 - zoom);
	}
```

When zoom = 0, `*x` would be ` ix * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31)`;
When zoom=32, `*x` also be ` ix * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31)`;

I don not understand why skip shift when zoom = 0? Maybe it is a bug. tiletoepsg3857 function also have some problem.

**Comment by ericfischer** [comment]
I think you're right. The special case for 0 is supposed to prevent trying to shift right by 32 bits, which is undefined behavior because it falls off the end of a 32-bit integer. But the result should be that `*x` and `*y` are 0, not their original values. I think this only works by coincidence, because this function only ever gets called with `zoom == 32`, so the zoom level 0 case is never exercised.

**Comment by jingsam** [comment]
Another question is y projection:

```
*y = ((1LL << 32) - 1) - (iy * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31));
```

I would expect:

```
*y = ((1LL << 32)) - (iy * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31));
```

Why *y offset by 1?

**Comment by jingsam** [comment]
One more question is about precision.
As *x is long long, so after:
```
*x = ix * (1LL << 31) / 6378137.0 / M_PI + (1LL << 31);
```

*x is rounded from double to long long.

After:
```
*x >>= (32 - zoom);
```

*x is also rounded.

My concern is that *x is rounded twice, maybe it cause precision loss?

**Comment by jingsam** [comment]
Fix above:
```
	*x = (ix / (2 * M_PI * 6378137.0) + 0.5) * (1LL << zoom);
	*y = (0.5 - iy / (2 * M_PI * 6378137.0)) * (1LL << zoom);
```

**Comment by ericfischer** [comment]
I agree that that the `-1` doesn't make sense either, and I don't remember why it was there in the first place, and also that it is better to multiply than to shift to avoid truncation. I'll try your implementation.

I haven't made a PR yet for this because I'm trying to make test cases for everything, and I'm not sure I understand yet what the correct behavior is for coordinates that are off the edge of the world at low zooms.
